core/src/forwarding_stage.rs

```diff
@@ -2,131 +2,177 @@
     agave_banking_stage_ingress_types::BankingPacketBatch,
     agave_transaction_view::transaction_view::SanitizedTransactionView,
+    async_trait::async_trait,
     crossbeam_channel::{Receiver, RecvTimeoutError},
     packet_container::PacketContainer,
     solana_client::connection_cache::ConnectionCache,
     solana_connection_cache::client_connection::ClientConnection,
     solana_cost_model::cost_model::CostModel,
-    solana_gossip::contact_info::Protocol,
+    solana_gossip::{cluster_info::ClusterInfo, contact_info::Protocol},
     solana_net_utils::bind_to_unspecified,
     solana_perf::data_budget::DataBudget,
     solana_poh::poh_recorder::PohRecorder,
     solana_runtime::{bank::Bank, root_bank_cache::RootBankCache},
     solana_runtime_transaction::{
         runtime_transaction::RuntimeTransaction, transaction_meta::StaticMeta,
     },
-    solana_sdk::{fee::FeeBudgetLimits, packet, transaction::MessageHash},
+    solana_sdk::{
+        fee::FeeBudgetLimits, packet, quic::NotifyKeyUpdate, signer::keypair::Keypair,
+        transaction::MessageHash,
+    },
     solana_streamer::sendmmsg::batch_send,
+    solana_tpu_client_next::{
+        connection_workers_scheduler::{ConnectionWorkersSchedulerConfig, Fanout},
+        leader_updater::LeaderUpdater,
+        transaction_batch::TransactionBatch,
+        ConnectionWorkersScheduler, ConnectionWorkersSchedulerError,
+    },
     std::{
-        net::{SocketAddr, UdpSocket},
-        sync::{Arc, RwLock},
+        net::{Ipv4Addr, SocketAddr, UdpSocket},
+        sync::{Arc, Mutex, RwLock},
         thread::{Builder, JoinHandle},
         time::{Duration, Instant},
     },
+    tokio::{runtime::Handle as RuntimeHandle, sync::mpsc, task::JoinHandle as TokioJoinHandle},
+    tokio_util::sync::CancellationToken,
 };
 
 mod packet_container;
 
+/// [`ForwardingClientOption`] enum represents the available client types for TPU
+/// communication:
+/// * [`ConnectionCacheClient`]: Uses a shared [`ConnectionCache`] to manage
+///       connections efficiently.
+/// * [`TpuClientNextClient`]: Relies on the `tpu-client-next` crate and
+///       requires a reference to a [`Keypair`].
+pub enum ForwardingClientOption<'a> {
+    ConnectionCache(Arc<ConnectionCache>),
+    TpuClientNext((&'a Keypair, RuntimeHandle)),
+}
+
 /// Value chosen because it was used historically, at some point
 /// was found to be optimal. If we need to improve performance
 /// this should be evaluated with new stage.
 const FORWARD_BATCH_SIZE: usize = 128;
 
-/// Addresses to use for forwarding.
 #[derive(Clone)]
-pub struct ForwardingAddresses {
-    /// Forwarding address for TPU (non-votes)
-    pub tpu: Option<SocketAddr>,
-    /// Forwarding address for votes
-    pub tpu_vote: Option<SocketAddr>,
+pub(crate) struct ForwardAddressGetter {
+    cluster_info: Arc<ClusterInfo>,
+    poh_recorder: Arc<RwLock<PohRecorder>>,
 }
 
-pub trait ForwardAddressGetter: Send + Sync + 'static {
```

Comment:
My understanding is that this trait is only used for test. I changed implementation of `ForwardingStage` to be parametrized by `ForwardingClient` (one for vote, one for non-vote) and, hence, can use `MockClient` in tests instead of using UDP sockets.

Is there any disagreement with this change or sounds good? @apfitzge

Reply:
Extracted this particular change in https://github.com/anza-xyz/agave/pull/5565